### PR TITLE
Light culling, Pixel culling

### PR DESCRIPTION
**Pull request by benanil**
we discard, spot and point lights with frustum culling and depth test now

### CODE REVIEW

File: ToolKit/MathUtil.h [hunk 1]
```diff
@@ -89,6 +89,20 @@ namespace ToolKit
...
+  
+  TK_API bool ConePointIntersection(Vec3 conePos, Vec3 coneDir, float coneHeight, float coneAngle, Vec3 point);
+  
+  TK_API bool FrustumConeIntersect(const Frustum& frustum, 
+                                   Vec3 conePos, 
+                                   Vec3 coneDir, 
+                                   float coneHeight, 
+                                   float coneAngle);
+
+  TK_API Quaternion QuaternionLookAt(Vec3 direction);
```
Comment (afraidofdark) [inline ToolKit/MathUtil.h:101]: Whats the assumption of the initial direction ? A quaternion define a rotation from dir a to dir b. I suppose direction is the target dir but what is the initial dir ?

File: ToolKit/MathUtil.cpp [hunk 2]
```diff
@@ -611,6 +611,59 @@ namespace ToolKit
     return res;
   }
 
+  Quaternion QuaternionLookAt(Vec3 direction)
+  {
+    glm::mat3x3 Result {};
```
Comment (afraidofdark) [inline ToolKit/MathUtil.cpp:616]: use Mat3 and fix the capilat Resoult -> result

File: ToolKit/AdditiveLightingPass.cpp [hunk 3]
```diff
@@ -155,21 +173,131 @@ namespace ToolKit
...
+
+      LightAndType() {}
+
+      LightAndType(Light* l, int t) : light(l), type(t) {}
+    };
+
+    std::vector<LightAndType> screenSpaceLights; // light and type pair
+    std::vector<std::pair<LightAndType, RenderJob>> meshLights;
+
     // Add all lights to a texture
```
Comment (afraidofdark) [inline ToolKit/AdditiveLightingPass.cpp:196]: old comment ? we don't use texture.

File: ToolKit/AdditiveLightingPass.cpp [hunk 4]
```diff
@@ -155,21 +173,131 @@ namespace ToolKit
...
+          // if we are inside of the cone we should draw on quad
+          screenSpaceLights.push_back({light, lightType});
+          continue;
+        }
+
+        if (!FrustumConeIntersect(frustum, pos, dir, height, outerAngle))
+        {
+          continue;// we don't see the cone, no need to render
+        }
+        uint64 hash = ConeMeshHash(height, outerAngle);
```
Comment (afraidofdark) [inline ToolKit/AdditiveLightingPass.cpp:251]: is there an upper bound for the cache ?  an artist can easly overflow gpu memory while editing spotlights.

Comment (afraidofdark) [reply ToolKit/AdditiveLightingPass.cpp:251]: what will be the removal algorithm ? frequency of access or number of referance ? If I were you I would create this model on each light and doesn't deal with all this trouble.

File: ToolKit/MathUtil.h (hunk 1 above)
Comment (benanil) [reply ToolKit/MathUtil.h:101]: If you think as EulerAngles initial direction is equal to vec3(0.0f)
you can think of it as: direction rotates the euler angles by direction * pi
so this function returns an quaternion that looks to direction. function can be used like this:
QuatLookAt(glm::normalize(FromPos - TargetPos))

**Review by afraidofdark — APPROVED** [review]

**Pull request merged**